src/main/java/rx/internal/operators/OperatorMerge.java

```diff
@@ -71,715 +72,878 @@
-                    if (emitted == r) {
-                        // we emitted as many as were requested so stop the forEach loop
-                        return Boolean.FALSE;
+                }
+                // in case there was some activity, continue with the regular emission loop
+                emitLoop();
+            } else {
+                // either someone else is emitting or there was no available request
+                // queue up the scalar
+                RxRingBuffer svq = scalarQueue;
+                if (svq == null) {
+                    svq = RxRingBuffer.getSpscInstance();
+                    add(svq);
+                    scalarQueue = svq;
+                }
+                try {
+                    svq.onNext(nl.next(value));
+                } catch (MissingBackpressureException ex) {
+                    onError(ex);
+                    return;
+                } catch (IllegalStateException ex) {
+                    // if there was an IAE and we are unsubscribed, just quit
+                    if (!isUnsubscribed()) {
+                        onError(ex);
                     }
+                    return;
                 }
-                return Boolean.TRUE;
-            }
-
-        };
-
-        @Override
-        public void onError(Throwable e) {
-            if (!completed) {
-                completed = true;
-                innerError(e, true);
+                // then try to emit it (or notify the running emission loop).
+                emit();
             }
         }
         
-        private void innerError(Throwable e, boolean parent) {
-            if (delayErrors) {
-                synchronized (this) {
-                    if (exceptions == null) {
-                        exceptions = new ConcurrentLinkedQueue<Throwable>();
+        void fastEmitNext(T value) {
+            synchronized (this) {
+                if (emitting) {
+                    int s = fastSize;
+                    Object[] o = fastArray;
+                    if (o == null) {
+                        o = new Object[4];
+                        fastArray = o;
+                    } else
+                    if (s == o.length) {
+                        Object[] o2 = new Object[s + (s >> 2)];
+                        System.arraycopy(o, 0, o2, 0, s);
+                        fastArray = o2;
+                        o = o2;
                     }
+                    o[s] = value;
+                    fastSize = s + 1;
+                    missed = true;
+                    return;
                 }
-                exceptions.add(e);
-                boolean sendOnComplete = false;
-                synchronized (this) {
-                    if (!parent) {
-                        wip--;
+                emitting = true;
+            }
+            boolean skipFinal = false;
+            try {
+                child.onNext(value);
+                for (int k = Integer.MAX_VALUE; k >= 0; k--) {
```

Comment:
Why the `MAX_VALUE` for-loop instead of `while(true)`?

In an extreme edge case of thread-starvation this seems like it could end the loop while there are still items in the queue. This would then mean the `child.onNext` on the previous line could emit while the queue is not empty for the next thread that comes along and picks it up ... or it could result in the system hanging if a terminal event was left in the queue.

It's an extreme edge case I'm sure, as `Integer.MAX_VALUE` is a large number, but I think it could theoretically happen.